yarn-project/sequencer-client/src/publisher/viem-tx-sender.ts

```diff
@@ -109,13 +125,32 @@ export class ViemTxSender implements L1PublisherTxSender {
         transactionHash: txHash,
         gasUsed: receipt.gasUsed,
         gasPrice: receipt.effectiveGasPrice,
+        logs: receipt.logs,
       };
     }
 
     this.log(`Receipt not found for tx hash ${txHash}`);
     return undefined;
   }
 
+  /**
+   * Publishes tx effects to Availability Oracle.
+   * @param encodedBody - Encoded block body.
+   * @returns The hash of the mined tx.
+   */
+  async sendPublishTx(encodedBody: Buffer): Promise<string | undefined> {
+    const args = [`0x${encodedBody.toString('hex')}`] as const;
+
+    const gas = await this.availabilityOracleContract.estimateGas.publish(args, {
+      account: this.account,
+    });
+    const hash = await this.availabilityOracleContract.write.publish(args, {
+      gas,
```

Comment:
In the real world should probably apply a small multiplier on the gas value here.

Reply:
Why is that so? Are the estimations in precise?

Reply:
If estimate is exact, and state that you touch is different than simulation you might revert. Don't know if the estimate is already applying one though.